src/shims/foreign_items.rs

```diff
@@ -591,14 +594,11 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
             }
             // underscore case for windows
             "_ldexp" | "ldexp" => {
-                // FIXME: Using host floats.
-                let x = f64::from_bits(this.read_scalar(args[0])?.to_u64()?);
+                let x = this.read_scalar(args[0])?.to_f64()?;
                 let exp = this.read_scalar(args[1])?.to_i32()?;
-                extern {
-                    fn ldexp(x: f64, n: i32) -> f64;
-                }
-                let n = unsafe { ldexp(x, exp) };
-                this.write_scalar(Scalar::from_u64(n.to_bits()), dest)?;
+                // For radix-2 (binary) systems, `ldexp` and `scalbn` are the same.
+                let res = x.scalbn(exp.try_into().unwrap());
```

Comment:
@eddyb the odd thing is that the C version supports `i32` but your version only `i16`. Panicking on big values seems wrong, any advise what else to do?

Reply:
You need to look at how LLVM constant-folds their libm intrinsics.

Reply:
Wait, I bet you can just saturate it to the largest `i16`.

Reply:
Based [on this snippet](https://github.com/rust-lang/rust/blob/be8bbb06976c8065425b18e9cbe24a6d1d4e7515/src/librustc_apfloat/ieee.rs#L1416-L1426) and the fact that `f64` uses `11` bits of exponent and `f128` would use `15` bits, you can safely clamp it to `ExpInt::min_value()..=ExpInt::max_value()` and it would still be outside of the valid exponent range (so the effect will be the same, presumably underflow to 0 / overflow to Inf).

Reply:
> Based on this snippet

So the only use of `exp` is in `exp as i32`? Seems more reasonable then to just expose it as an `i32`, IMO.

Reply:
But for now I went with clamping, thanks.

Reply:
`i32` is only used as an intermediary to avoid overflow (with `f128`, it could stay entirely within `i16` for `f64`).
Most storage and manipulation of the exponent is done in `ExpInt`, which is `i16`.

Changing `ExpInt` to `i32` would be possible (although wasteful IMO), it would just take more effort. And I'd prefer not to mess with this one function (which I only ported for completeness sake, I didn't expect it to be useful).